src/ops/logical.rs

```diff
@@ -0,0 +1,554 @@
+use crate::array::Array;
+use crate::stream::StreamOrDevice;
+use mlx_macros::default_device;
+
+impl Array {
+    /// Unary element-wise logical not.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use mlx::Array;
+    /// let a: Array = false.into();
+    /// let mut b = a.logical_not_device(Default::default());
+    ///
+    /// b.eval();
+    /// let b_data: &[bool] = b.as_slice();
+    /// // b_data == [true]
+    /// ```
+    ///
+    /// # Params
+    ///
+    /// - stream: stream or device to evaluate on
+    #[default_device]
+    pub fn logical_not_device(&self, stream: StreamOrDevice) -> Array {
+        unsafe { Array::from_ptr(mlx_sys::mlx_logical_not(self.c_array, stream.as_ptr())) }
+    }
+
+    /// Element-wise equality.
+    ///
+    /// Equality comparison on two arrays with <doc:broadcasting>.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use mlx::Array;
+    /// let a = Array::from_slice(&[1, 2, 3], &[3]);
+    /// let b = Array::from_slice(&[1, 2, 3], &[3]);
+    /// let mut c = a.eq_device(&b, Default::default());
+    ///
+    /// c.eval();
+    /// let c_data: &[bool] = c.as_slice();
+    /// // c_data == [true, true, true]
+    /// ```
+    ///
+    /// # Params
+    ///
+    /// - other: array to compare
+    /// - stream: stream or device to evaluate on
+    #[default_device]
+    pub fn eq_device(&self, other: &Array, stream: StreamOrDevice) -> Array {
```

Comment:
What if two arrays are of different shape? What about doing the `*_unchecked` and `try_*` here?

Reply:
The c++ code checks whether the two shapes are broadcastable